pylint, pep8 related issues fixed in pylti

amir-qayyum-arbisoft opened the pull request:
Hi

In this PR I fixed all issues diagnose using pep8 and pylint. @carsongee I took pylintrc file from edx-plateform and placed here with little modification. Can you review it?

Thanks
#33

carsongee commented on pylti/tests/test_flask_app.py:
```diff
@@ -1,32 +1,50 @@
+        """
         self.exception = None
 
-app_exception = ExceptionHandler()
```
I would actually just do the invalid-name exception here since a function really isn't a constant

carsongee commented:
This will need a rebase, and I'd also prefer not to have a pylintrc at all unless we find reason to need one.  If we do, I'd rather we start with the default one: `pylint --generate-rcfile` and modify it to fit our needs rather than copying the [edx-lint](https://github.com/edx/edx-lint) generated one from edx-platform.

amir-qayyum-arbisoft commented:
@carsongee  done with rebase and generating pylintrc file.

carsongee commented:
I'm still seeing pylint errors:
```
************* Module pylti.common
W: 51, 0: Method 'authorize_request_token' is abstract in class 'OAuthDataStore' but is not overridden (abstract-method)
W: 51, 0: Method 'fetch_access_token' is abstract in class 'OAuthDataStore' but is not overridden (abstract-method)
W: 51, 0: Method 'fetch_request_token' is abstract in class 'OAuthDataStore' but is not overridden (abstract-method)
W: 51, 0: Method 'lookup_token' is abstract in class 'OAuthDataStore' but is not overridden (abstract-method)
```
These are all safe to disable under the class declaration for `LTIOAuthDataStore` since they are unused.  Also, there is no reason to check the pylintrc file into git since the default configuration is what we want for now.  So go ahead and just remove that file altogether.  With those two things we should be good to go.  You probably also want to make sure you have the latest version of pylint as they are always coming up with new things to check for.  I tested this with 1.4.3

amir-qayyum-arbisoft commented:
@carsongee in the pylintrc file I commit it is generated plus customized. I have added a list of warning messages in disable option. So If I remove this file I need to put each disable message into code 
**disbale = { list of warning messages}**

Second I am using pylint with version 1.0.1 it is the most stable version. I am on linux I installed it by following command
*sudo apt-get install pylint*

carsongee commented:
You should just be able to install pip, and run `pip install --upgrade pylint`, and as I mentioned, the messages didn't change at all with and without the RC file, so I don't think it is actually doing anything right now.  The default configuration file seems to be just fine for what we have here other than the message I posted which should be addressed locally, and not with the rcfile.

amir-qayyum-arbisoft commented:
@carsongee i just updated pylint to 1.4.3, I am not able to reproduce warnings.. screenshots are attached 
![screen shot 2015-04-21 at 5 42 20 pm](https://cloud.githubusercontent.com/assets/10431250/7252434/6b16d39c-e84e-11e4-9479-c3afae608138.png)
![screen shot 2015-04-21 at 5 39 36 pm](https://cloud.githubusercontent.com/assets/10431250/7252436/6f953c88-e84e-11e4-8557-e0e10dda63e9.png)

carsongee commented:
I don't know why you aren't seeing these.  The only thing I can think of is that pylint isn't seeing your virtual environment correctly and thus doesn't know that you are aren't implementing all the abstract methods.   Either way, the fix is easy by adding `# pylint: disable=abstract-method` to line 56 in `common.py`

carsongee commented on pylintrc:
```diff
@@ -0,0 +1,407 @@
+        logging-format-interpolation,
+        import-error,
+	unused-argument,
+        unused-variable,
```
I'm really not comfortable with almost any of these checks being disabled.  We really shouldn't have unused-variables in our code or really most of these.  The most notable ones I actually want us to correct with this PR are `unused-variable`, `import-error`, `logging-format-interpolation`, `no-self-use`, `redundant-unittest-assert`, `redifined-outer-name`, `unused-argument`, and `duplicate-code`.  For other more stylistic ones like `too-many-locals`, or `invalid-name` we can possibly deal with these in our RC file by changing properties of the checks like allowing longer method names, or more local variables, but we shouldn't be disabling them all together.  I really think you should try and address all the pylint issues as they are with just the default RC file and use local disable statements. The only good changes to the RC file that make sense to me are increasing the allowed names for methods.

amir-qayyum-arbisoft replied:
@carsongee  sure, let me review code again and fix issues according to your suggestions.

Thanks,

carsongee commented on pylti/flask.py:
```diff
@@ -106,7 +106,7 @@ def _verify_any(self):
             self.verify_request()
 
-    def _verify_session(self):
+    def _verify_session(self):  # pylint: disable=no-self-use
```
if this doesn't have self use, you should decorate it with `@staticmethod` and remove self from the arguments

carsongee commented on pylti/flask.py:
```diff
@@ -68,7 +68,7 @@ def name(self):
 
     @property
-    def user_id(self):
+    def user_id(self):  # pylint: disable=no-self-use
```
for the `@property` methods, I think you are stuck using self anyway, so these disables are ok

carsongee commented on pylti/common.py:
```diff
@@ -243,6 +245,7 @@ def post_message2(consumers, lti_key, url, body,
     :param: body: xml body
     :return: success
     """
+    # pylint: disable=unused-variable,too-many-arguments
```
You should actually remove whatever variable is unused

carsongee commented on pylti/common.py:
```diff
@@ -216,6 +217,7 @@ def post_message(consumers, lti_key, url, body):
     :param body: xml body
     :return: success
     """
+    # pylint: disable=unused-variable
```
again, you should actually remove, or replace with an _ any variable that is unused

carsongee commented on pylti/common.py:
```diff
@@ -49,11 +49,12 @@
     for app engine at https://code.google.com/p/ims-dev/
     """
-
+    # pylint: disable=bad-option-value
```
why is that disabled, and why are we disabling in two different places?

carsongee commented on pylti/flask.py:
```diff
@@ -160,7 +160,7 @@ def role(self):
         return session['roles']
 
-    def is_role(self, role):
+    def is_role(self, role):  # pylint: disable=no-self-use
```
remove self and add `@staticmethod`

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -52,10 +50,14 @@ class TestCommon(unittest.TestCase):
 
     def test_version(self):
-        # Will raise ValueError if not a semantic version
+        # pylint: disable=no-self-use
```
another one for `@staticmethod`

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -52,10 +50,14 @@ class TestCommon(unittest.TestCase):
 
-    def test_ltioauthdatastore(self):
+    def test_lti_oauth_data_store(self):
+        # pylint: disable=invalid-name
```
why is this an invalid name?

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -89,33 +92,40 @@ def test_verify_request_common(self):
         """
         verify_request_common succeeds on valid request
         """
+        # pylint: disable=unused-variable
```
again, this should just be fixed so we don't use the variable

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -76,7 +78,8 @@ def test_ltioauthdatastore(self):
 
-    def test_ltioauthdatastore_no_consumers(self):
+    def test_lti_oauth_data_store_no_consumers(self):
+        # pylint: disable=invalid-name
```
So I think for these long names we can just create a pylintrc file that allows for longer names.  Thus far the entire contents of the rc file would be:
```ini
[BASIC]

# Regular expression which should only match correct method names
method-rgx=[a-z_][a-z0-9_]{2,40}$
```
Where you will need to adjust the `40` in `{2,40}` for the right length of method names as longs ones are fine

amir-qayyum-arbisoft replied:
@carsongee done with it. but pylint is still giving me invalid-name error, so i disable this messages for these specific functions

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -89,33 +92,40 @@ def test_verify_request_common(self):
         """
         verify_request_common succeeds on valid request via proxy
         """
+        # pylint: disable=unused-variable
```
same

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -89,33 +92,40 @@ def test_verify_request_common(self):
         """
         verify_request_common fails on missing authentication
         """
+        # pylint: disable=unused-variable, invalid-name
```
same

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -216,6 +235,7 @@ def test_generate_xml(self):
 </imsx_POXBody></imsx_POXEnvelopeRequest>""")
 
     def generate_oauth_request(self, url_to_sign=None):
+        # pylint: disable=no-self-use
```
`@staticmethod`

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -73,6 +74,7 @@ def get_exception(self):
         return app_exception.get()
 
     def has_exception(self):
+        # pylint: disable=no-self-use
```
`@staticmethod`

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -173,9 +182,11 @@ def test_access_to_oauth_resource(self):
         """
         Check that name is returned if passed via initial request
         """
+        # pylint: disable=maybe-no-member
```
I don't know what this message is, but you should try to fix the cause instead of just disabling the message.  Also, you have disables at the same scope in two different places

amir-qayyum-arbisoft replied:
@carsongee  fix this one , I made response iterate-able by using yield. Otherwise i was not able to access dict of response because it was not iterate-able. No pyltint is not giving maybe-no-member messages

carsongee commented on pylti/tests/test_flask_app.py:
```diff
@@ -1,38 +1,56 @@
 
 @app.route("/unknown_protection")
-@lti(error=error, app=app, request='notreal')
+@lti_flask(error=error, app=app, request='notreal')
```
why did you change the decorator name?

amir-qayyum-arbisoft replied:
@carsongee because the name of decorator and the param was same. I was getting a warning from pylint which state that I am modifying outer variable. On changing the param of function i got error from test_flash.py test start getting fail

carsongee commented:
You'll need to figure out how to get the coverage back to 100%.  I don't know why half of the tests don't appear to be getting run, but you'll definitely want to figure that out.

amir-qayyum-arbisoft commented:
@carsongee  I am working on it

amir-qayyum-arbisoft commented:
@carsongee i am working on this PR, please give me some time 

Thanks

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -173,9 +183,11 @@ def test_access_to_oauth_resource(self):
         self.assertFalse(self.has_exception())
 
     def test_access_to_oauth_resource_name_passed(self):
+        # pylint: disable=invalid-name
```
I think you are still getting these because you have allowed longer function names, but these are methods, not functions. Again, I think the entire contents of your pylintrc file should be:
```ini
[BASIC]

# Regular expression which should only match correct method names
method-rgx=[a-z_][a-z0-9_]{2,40}$
```

and possibly
```ini
[BASIC]

# Regular expression which should only match correct method names
method-rgx=[a-z_][a-z0-9_]{2,40}$

# Regular expression matching correct function names
function-rgx=[a-z_][a-z0-9_]{2,40}$
```
All of that extra stuff in that file is default, and thus not needed.

amir-qayyum-arbisoft commented:
@carsongee done with changes and fixed coverage issues,

Added **logging-format-interpolation** in RC file because i think .format() is ok to use. 
 http://stackoverflow.com/questions/5082452/python-string-formatting-vs-format

their are some messages like multiple arguments, multiple params , I thing messages like it should be add into RC file. What do you say?

carsongee commented:
string format is better for everything but logging, which is what that pylint message is telling you.  For logging, the format should be `log.debug('Hello %s, how are you %s', name, 'today')` otherwise you are making an extra function call that is not needed.

carsongee commented on pylti/flask.py:
```diff
@@ -192,7 +195,7 @@ def _check_role(self):
         log.debug("check_role lti_role={} decorator_role={}"
                   .format(self.role, role))
-        if not self.is_role(role):
+        if not LTI.is_role(role):
```
You should still use `self.is_role` here since it isn't a class method, but a static method, and if we ever needed to add self, we wouldn't have to change these calls.

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -65,30 +66,33 @@ def setUp(self):
         :return: text
         """
-        return "{}".format(self.get_exception())
+        return "{}".format(TestFlask.get_exception())
```
same here, still use self.

carsongee commented on pylti/flask.py:
```diff
@@ -181,7 +184,7 @@ def is_role(self, role):
             raise LTIException("Unknown role {}.".format(role))
 
-    def _check_role(self):
+    def check_role(self):
```
I think this should keep it's name `_check_role` and we should add back the pylint disable since this isn't ever meant to be exposed to users.

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -17,15 +17,16 @@
     """
     Valid XML response from LTI 1.0 consumer
-    """
+    """  # pylint: disable=pointless-string-statement
```
You should actually just remove this, or convert to a comment

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
 
         self.app.get('/session')
-        self.assertFalse(self.has_exception())
+        self.assertFalse(TestFlask.has_exception())
```
should still be self

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
         self.app.get('/session')
 
-        self.assertFalse(self.has_exception())
+        self.assertFalse(TestFlask.has_exception())
```
should still be self

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
         self.app.get('/session')
 
-        self.assertTrue(self.has_exception())
+        self.assertTrue(TestFlask.has_exception())
```
self here

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
         consumers = self.consumers
         url = 'http://localhost/initial?'
-        new_url = self.generate_launch_request(consumers, url)
+        new_url = TestFlask.generate_launch_request(consumers, url)
```
should be self

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
 
         self.app.get(new_url)
-        self.assertFalse(self.has_exception())
+        self.assertFalse(TestFlask.has_exception())
```
self again.

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -138,131 +142,139 @@ def test_access_to_oauth_resource_without_authorization_initial_post(self):
 
         ret = self.app.get(new_url)
-        self.assertFalse(self.has_exception())
+        self.assertFalse(TestFlask.has_exception())
```
I think you get the picture :)

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -103,25 +105,27 @@ def test_verify_request_common_via_proxy(self):
         try:
             ret = verify_request_common(consumers, url, method,
                                         headers, params)
         except LTIException:
```
These should actually be:
```python
with self.assertRaises(LTIException):
    verify_request_common(consumers, url, method,
                                         headers, params)
```
Then you don't need `ret` or the self.assertFalse

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -140,7 +144,7 @@ def test_verify_request_common_no_params(self):
             ret = verify_request_common(consumers, url, method,
                                         headers, params)
         except LTIException:
```
Same here:
```python
with self.assertRaises(LTIException):
    verify_request_common(consumers, url, method,
                                         headers, params)
```
Then you don't need `ret` or the self.assertFalse

carsongee commented:
Just to address your other comment, for multiple arguments and multiple params, those shouldn't be added to the RC file as we should strive to write functions that don't need that many arguments.  We are stuck for the most part right now with what we have, but if we added a new function tomorrow, I would want it to fit into the pylint defaults.  That is the reason for just doing local disable messages there.

amir-qayyum-arbisoft commented:
@carsongee done with all changes on your feedback. Now it is giving me code-duplication warning between 2 classes test_flask and test_common , is this ok?

carsongee commented on pylti/flask.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 """
-    PyLTI decorator implementation for flask framework
+    PyfLTI decorator implementation for flask framework
```
This should still be PyLTI

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -100,31 +104,32 @@ def test_verify_request_common_via_proxy(self):
         """
         verify_request_common succeeds on valid request via proxy
         """
+        # pylint: disable= invalid-name
```
remove

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -100,31 +104,32 @@ def test_verify_request_common_via_proxy(self):
         """
         verify_request_common fails on missing authentication
         """
+        # pylint: disable= invalid-name
```
remove

carsongee commented on pylti/tests/test_common.py:
```diff
@@ -100,31 +104,32 @@ def test_verify_request_common_via_proxy(self):
+            verify_request_common(consumers, url, method, headers, params)
 
     def test_verify_request_common_no_params(self):
+        # pylint: disable=invalid-name
```
remove

carsongee commented on pylti/tests/test_flask.py:
```diff
@@ -65,30 +64,34 @@ def setUp(self):
+        return "{}".format(TestFlask.get_exception())
 
     def test_access_to_oauth_resource_unknown_protection(self):
+        # pylint: disable=invalid-name
```
I tried removing all of these and there were no vioations.  Please remove all of the invalid-name disables.

amir-qayyum-arbisoft commented:
@carsongee done with removing invalid-name and fixing issues, Thanks

carsongee commented:
excellent, this looks good. :birthday:
